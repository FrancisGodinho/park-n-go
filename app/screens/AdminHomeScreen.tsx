import {
  Keyboard,
  SafeAreaView,
  StyleSheet,
  Text,
  TouchableWithoutFeedback,
  View,
} from "react-native";
import React, { useEffect, useState } from "react";
import { useFormik } from "formik";
import * as Yup from "yup";
import { auth, db } from "../utils/Firebase";
import { doc, getDoc, setDoc, updateDoc } from "firebase/firestore";
import ParkingAPI from "../api/ParkingAPI";
import Headers from "../constants/Headers";
import Colors from "../constants/Colors";
import { useGlobalContext } from "../utils/context";
import CustomButton from "../components/CustomButton";
import CustomTextInput from "../components/CustomTextInput";
import MapView, { Marker } from "react-native-maps";

const LOT_ID = "jw7d1mNE2Cw1mTG0tzzH";
type Props = {};

// const lotSchema = Yup.object().shape({
//   rate: Yup.number()
//     .required()
//     .test("something", "LOL", (value: number) => {
//       console.log("value: " + value);
//       return value > 0;
//     }),
// });

const AdminHomeScreen = (props: Props) => {
  const [lotRate, setLotRate] = useState("");
  const [lotCapacity, setLotCapacity] = useState("");
  const [curNumCars, setCurNumCars] = useState(0);
  const [lotLongitude, setLotLongitude] = useState("");
  const [lotLatitude, setLotLatitude] = useState("");
  const [lotCoordinate, setLotCoordinate] = useState({
    coordinate: { longitude: 0, latitude: 0 },
  });
  const [lotDescription, setLotDescription] = useState("");

  // Note that lotId is null because the admin was not associated with any parking lot during registration
  // In reality an admin should contact us to set up the parkade for security reasons
  // Currently using a fixed LOT_ID for this admin
  const { lotName, lotId, setLotName } = useGlobalContext();

  const formik = useFormik({
    initialValues: {
      rate: lotRate,
      capacity: lotCapacity,
      longitude: lotLongitude,
      latitude: lotLatitude,
    },
    enableReinitialize: true,
    // validationSchema: lotSchema,
    onSubmit: (values) => {
      console.log(values);
    },
  });

  useEffect(() => {
    (async () => {
      const lotSnap = await getDoc(doc(db, "lots", LOT_ID));
      setLotName(lotSnap.data()?.name);
      setLotRate(lotSnap.data()?.rate);
      setLotCapacity(lotSnap.data()?.capacity);
      setCurNumCars(lotSnap.data()?.currentNumCars);
      setLotLongitude(lotSnap.data()?.longitude);
      setLotLatitude(lotSnap.data()?.latitude);
      setLotCoordinate({
        coordinate: {
          longitude: parseFloat(lotSnap.data()?.longitude),
          latitude: parseFloat(lotSnap.data()?.latitude),
        },
      });
      setLotDescription(`Rate: $${lotSnap.data()?.rate} per hour`);
    })();
  }, [LOT_ID]);

  const signOut = () => {
    auth.signOut();
  };

  const updateRate = (newRate: string) => {
    setLotRate(newRate);
    (async () => {
      await updateDoc(
        doc(db, "lots", LOT_ID),
        "rate",
        parseFloat(formik.values.rate)
      );
    })();
  };

  const updateCapacity = (newCapacity: string) => {
    setLotCapacity(newCapacity);
    (async () => {
      await updateDoc(
        doc(db, "lots", LOT_ID),
        "capacity",
        parseFloat(formik.values.capacity)
      );
    })();
  };

  const updateLongitude = (newLongitude: string) => {
    setLotLongitude(newLongitude);
    (async () => {
      await updateDoc(
        doc(db, "lots", LOT_ID),
        "longitude",
        parseFloat(formik.values.longitude)
      );
    })();
  };

  const updateLatitude = (newLatitude: string) => {
    setLotLatitude(newLatitude);
    (async () => {
      await updateDoc(
        doc(db, "lots", LOT_ID),
        "latitude",
        parseFloat(formik.values.latitude)
      );
    })();
  };

  return (
    <TouchableWithoutFeedback onPress={Keyboard.dismiss} accessible={false}>
      <SafeAreaView testID="admin-safe-view" style={styles.pageContainer}>
        <Text style={[Headers.h1, styles.h1]}>{lotName}</Text>
        <View style={styles.pageContainer}>
          <View style={styles.itemContainer}>
            <View style={styles.itemTextContainer}>
              <Text style={styles.itemText} testID="rate-text">
                {"Rate: "}
              </Text>
              <Text style={[Headers.h2, styles.itemTextBody]}>
                ${lotRate}/hr
              </Text>
            </View>
            <CustomTextInput
              formik={formik}
              name="rate"
              style={styles.customInput}
              placeholder={""}
              value={formik.values.rate}
              onSubmitEditing={() => {
                updateRate(formik.values.rate);
              }}
            />
          </View>
          <View style={styles.itemContainer}>
            <View style={styles.itemTextContainer}>
              <Text style={styles.itemText} testID="capacity-text">
                {"Capacity: "}
              </Text>
              <Text style={[Headers.h2, styles.itemTextBody]}>
                {curNumCars}/{lotCapacity}
              </Text>
            </View>
            <CustomTextInput
              formik={formik}
              name="capacity"
              style={styles.customInput}
              placeholder={""}
              value={formik.values.capacity}
              onSubmitEditing={() => {
                updateCapacity(formik.values.capacity);
              }}
            />
          </View>
          <View style={styles.itemContainer}>
            <View style={styles.itemTextContainer}>
              <Text style={styles.itemText} testID="longitude-text">
                {"Longitude: "}
              </Text>
              <Text style={[Headers.h2, styles.itemTextBody]}>
                {lotLongitude}
              </Text>
            </View>
            <CustomTextInput
              formik={formik}
              name="longitude"
              style={styles.customInput}
              placeholder={""}
              value={formik.values.longitude}
              onSubmitEditing={() => {
                updateLongitude(formik.values.longitude);
              }}
            />
          </View>
          <View style={{ marginBottom: 10, ...styles.itemContainer }}>
            <View style={styles.itemTextContainer}>
              <Text style={styles.itemText} testID="latitude-text">
                {"Latitude: "}
              </Text>
              <Text style={[Headers.h2, styles.itemTextBody]}>
                {lotLatitude}
              </Text>
            </View>
            <CustomTextInput
              editable
              formik={formik}
              name="latitude"
              style={styles.customInput}
              placeholder={""}
              value={formik.values.latitude}
              onSubmitEditing={() => {
                updateLatitude(formik.values.latitude);
              }}
            />
          </View>
          <MapView style={styles.mapView} showsUserLocation>
            <Marker
              coordinate={lotCoordinate.coordinate}
              title={lotName}
              description={lotDescription}
            />
          </MapView>
          <CustomButton disabled={false} text="Sign Out" onPress={signOut} />
        </View>
      </SafeAreaView>
    </TouchableWithoutFeedback>
  );
};

export default AdminHomeScreen;

const styles = StyleSheet.create({
  pageContainer: {
    flex: 1,
    paddingHorizontal: 10,
    borderTopLeftRadius: 15,
    borderTopRightRadius: 15,
    backgroundColor: Colors.black,
  },
  h1: { marginTop: 10 },
  itemContainer: {
    width: "100%",
    justifyContent: "center",
    alignItems: "center",
    marginTop: 5,
  },
  itemTextContainer: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "flex-start",
  },
  itemText: {
    color: Colors.white,
    fontWeight: "bold",
  },
  itemTextBody: { color: Colors.primary },
  customInput: {
    maxWidth: "90%",
    maxHeight: 15,
  },
  signOutButton: {
    maxWidth: "30%",
  },
  mapView: {
    height: "40%",
  },
});
