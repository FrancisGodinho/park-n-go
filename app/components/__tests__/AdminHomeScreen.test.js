import React from "react";
import renderer from "react-test-renderer";
import { cleanup, render } from "@testing-library/react-native";
import { fireEvent } from "@testing-library/react";

import AdminHomeScreen from "../../screens/AdminHomeScreen";

beforeAll((done) => {
  done();
});

afterAll((done) => {
  done();
});

let findTextElement = function (tree, element) {
  console.warn(tree);
  return true;
};

describe("<AdminHomeScreen />", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<AdminHomeScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it.skip("rate text element exists", () => {
    const { getByTestId } = render(<AdminHomeScreen />);
    const rate = getByTestId("rate-text");
    expect(rate).toBeTruthy();
  });

  it.skip("capacity text element exists", () => {
    const { getByTestId } = render(<AdminHomeScreen />);
    const capacity = getByTestId("capacity-text");
    expect(capacity).toBeTruthy();
  });

  it.skip("longitude text element exists", () => {
    const { getByTestId } = render(<AdminHomeScreen />);
    const longitude = getByTestId("longitude-text");
    expect(longitude).toBeTruthy();
  });

  it.skip("latitude text element exists", () => {
    const { getByTestId } = render(<AdminHomeScreen />);
    const latitude = getByTestId("longitude-text");
    expect(latitude).toBeTruthy();
  });
});
