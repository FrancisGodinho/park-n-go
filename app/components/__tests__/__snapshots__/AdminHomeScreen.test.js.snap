// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AdminHomeScreen /> renders correctly 1`] = `
<RCTSafeAreaView
  accessible={false}
  emulateUnlessSupported={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#202020",
      "borderTopLeftRadius": 15,
      "borderTopRightRadius": 15,
      "flex": 1,
      "paddingHorizontal": 10,
    }
  }
  testID="admin-safe-view"
>
  <Text
    style={
      Array [
        Object {
          "color": "#C0C0C0",
          "fontFamily": "Avenir-Black",
          "fontSize": 25,
          "fontWeight": "900",
          "textAlign": "center",
        },
        Object {
          "marginTop": 10,
        },
      ]
    }
  >
    
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "#202020",
        "borderTopLeftRadius": 15,
        "borderTopRightRadius": 15,
        "flex": 1,
        "paddingHorizontal": 10,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginTop": 5,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#E0E0E0",
              "fontWeight": "bold",
            }
          }
          testID="rate-text"
        >
          Rate: 
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#C0C0C0",
                "fontFamily": "Avenir-Black",
                "fontSize": 20,
                "fontWeight": "700",
              },
              Object {
                "color": "#66A182",
              },
            ]
          }
        >
          $
          
          /hr
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#404040",
            "borderRadius": 5,
            "flexDirection": "row",
            "marginVertical": 10,
            "paddingVertical": 10,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCorrect={false}
          editable={true}
          enablesReturnKeyAutomatically={true}
          keyboardType="default"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder=""
          placeholderTextColor="#A0A0A0"
          rejectResponderTermination={true}
          returnKeyType="next"
          style={
            Array [
              Object {
                "color": "#C0C0C0",
                "fontFamily": "Avenir-Black",
                "fontSize": 17,
                "fontWeight": "400",
              },
              Object {
                "flex": 1,
                "paddingHorizontal": 10,
              },
              Object {
                "maxHeight": 15,
                "maxWidth": "90%",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginTop": 5,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#E0E0E0",
              "fontWeight": "bold",
            }
          }
          testID="capacity-text"
        >
          Capacity: 
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#C0C0C0",
                "fontFamily": "Avenir-Black",
                "fontSize": 20,
                "fontWeight": "700",
              },
              Object {
                "color": "#66A182",
              },
            ]
          }
        >
          0
          /
          
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#404040",
            "borderRadius": 5,
            "flexDirection": "row",
            "marginVertical": 10,
            "paddingVertical": 10,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCorrect={false}
          editable={true}
          enablesReturnKeyAutomatically={true}
          keyboardType="default"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder=""
          placeholderTextColor="#A0A0A0"
          rejectResponderTermination={true}
          returnKeyType="next"
          style={
            Array [
              Object {
                "color": "#C0C0C0",
                "fontFamily": "Avenir-Black",
                "fontSize": 17,
                "fontWeight": "400",
              },
              Object {
                "flex": 1,
                "paddingHorizontal": 10,
              },
              Object {
                "maxHeight": 15,
                "maxWidth": "90%",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginTop": 5,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#E0E0E0",
              "fontWeight": "bold",
            }
          }
          testID="longitude-text"
        >
          Longitude: 
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#C0C0C0",
                "fontFamily": "Avenir-Black",
                "fontSize": 20,
                "fontWeight": "700",
              },
              Object {
                "color": "#66A182",
              },
            ]
          }
        >
          
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#404040",
            "borderRadius": 5,
            "flexDirection": "row",
            "marginVertical": 10,
            "paddingVertical": 10,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCorrect={false}
          editable={true}
          enablesReturnKeyAutomatically={true}
          keyboardType="default"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder=""
          placeholderTextColor="#A0A0A0"
          rejectResponderTermination={true}
          returnKeyType="next"
          style={
            Array [
              Object {
                "color": "#C0C0C0",
                "fontFamily": "Avenir-Black",
                "fontSize": 17,
                "fontWeight": "400",
              },
              Object {
                "flex": 1,
                "paddingHorizontal": 10,
              },
              Object {
                "maxHeight": 15,
                "maxWidth": "90%",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginBottom": 10,
          "marginTop": 5,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#E0E0E0",
              "fontWeight": "bold",
            }
          }
          testID="latitude-text"
        >
          Latitude: 
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#C0C0C0",
                "fontFamily": "Avenir-Black",
                "fontSize": 20,
                "fontWeight": "700",
              },
              Object {
                "color": "#66A182",
              },
            ]
          }
        >
          
        </Text>
      </View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#404040",
            "borderRadius": 5,
            "flexDirection": "row",
            "marginVertical": 10,
            "paddingVertical": 10,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          autoCorrect={false}
          editable={true}
          enablesReturnKeyAutomatically={true}
          keyboardType="default"
          onChangeText={[Function]}
          onSubmitEditing={[Function]}
          placeholder=""
          placeholderTextColor="#A0A0A0"
          rejectResponderTermination={true}
          returnKeyType="next"
          style={
            Array [
              Object {
                "color": "#C0C0C0",
                "fontFamily": "Avenir-Black",
                "fontSize": 17,
                "fontWeight": "400",
              },
              Object {
                "flex": 1,
                "paddingHorizontal": 10,
              },
              Object {
                "maxHeight": 15,
                "maxWidth": "90%",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <AIRMap
      handlePanDrag={false}
      initialRegion={null}
      onChange={[Function]}
      onLayout={[Function]}
      onMapReady={[Function]}
      onMarkerPress={[Function]}
      region={null}
      showsUserLocation={true}
      style={
        Object {
          "height": "40%",
        }
      }
    >
      <AIRMapMarker
        coordinate={
          Object {
            "latitude": 0,
            "longitude": 0,
          }
        }
        description=""
        onPress={[Function]}
        stopPropagation={false}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "position": "absolute",
            },
            undefined,
          ]
        }
        title=""
      />
    </AIRMap>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#66A182",
          "borderRadius": 5,
          "marginBottom": 10,
          "marginTop": "auto",
          "opacity": 1,
          "paddingVertical": 10,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#C0C0C0",
              "fontFamily": "Avenir-Black",
              "fontSize": 17,
              "fontWeight": "400",
            },
            Object {
              "color": "#E0E0E0",
              "textAlign": "center",
            },
            false,
          ]
        }
      >
        Sign Out
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
